ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java

```diff
@@ -347,30 +349,41 @@ private static UnsignedLong get_base_reward(BeaconState state, int index) {
     }
 
     // Proposer and inclusion delay micro-rewards
-    for (Integer index : get_unslashed_attesting_indices(state, matching_source_attestations)) {
-      matching_source_attestations.stream()
-          .filter(
-              a ->
-                  get_attesting_indices(state, a.getData(), a.getAggregation_bits())
-                      .contains(index))
-          .min(Comparator.comparing(PendingAttestation::getInclusion_delay))
-          .ifPresent(
-              attestation -> {
-                UnsignedLong proposer_reward =
-                    get_base_reward(state, index)
-                        .dividedBy(UnsignedLong.valueOf(PROPOSER_REWARD_QUOTIENT));
-                rewards.set(
-                    attestation.getProposer_index().intValue(),
-                    rewards.get(attestation.getProposer_index().intValue()).plus(proposer_reward));
-                UnsignedLong max_attester_reward =
-                    get_base_reward(state, index).minus(proposer_reward);
-                rewards.set(
-                    index,
-                    rewards
-                        .get(index)
-                        .plus(max_attester_reward.dividedBy(attestation.getInclusion_delay())));
-              });
-    }
+
+    // map (unslashed attester index) -> (list of source attestations)
+    Map<Integer, List<PendingAttestation>> validator_source_attestations =
+        matching_source_attestations.stream()
+            .flatMap(
+                a ->
+                    get_unslashed_attesting_indices(state, Collections.singletonList(a)).stream()
```

Comment:
Do you mind explaining why you call the get_unslashed_attesting_indices here with single attestations instead of all of them together? I'm worried that this might be straying from the spec definition?

Reply:
The original spec definition lacks performance in favor of simplicity, so the heavier optimizations we would like to apply the greater changes we would need to do in the straightforward spec implementation. 

> why you call the get_unslashed_attesting_indices here with single attestations instead of all of them together?

I'm building a map `(validator) -> (his attestations)` instead of doing it via 3 nested loops as in original spec. I bet this change keeps the implementation compatible with the spec, however regression may always occur. 

You may check the initial fix I did: https://github.com/Nashatyrev/artemis/pull/4 It is a bit slower than this one but is much closer to original spec.

Reply:
It takes a bit of time to get your head around what both versions are actually doing but they are equivalent.

The spec version finds all the validators (by index) included in any of the `matching_source_attestations`, then for each validator finds the list of attestations they were included in.
The new version builds a map of validator to attestations they were included in by iterating each attestation from `matching_source_attestations` and converting to a pair of (validator, attestation) then grouping them by validator (lines 359 to 361).

The majority of the method is unchanged except for indenting, for each validator find the attestation with minimum inclusion delay and apply rewards based on that (lines 355-372 in the old and 363-385 in the new).

Reply:
@ajsutton yes, exactly. Thanks for detailed description!

Reply:
Thanks guys!